### ISSUE TITLE
The osd pod failed randomly when creating 2 rook-ceph clusters on the same set of storage nodes

### ISSUE THREAD

**Issue by yan636** [rook/rook#2696]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
One of two osd pods failed randomly. 

**Expected behavior:**
All osd pods should be in "Running" state consistently.

**How to reproduce it (minimal and precise):**

1). Failure case:  Two clusters with one cluster with 1 device 'vdd' and the other cluster with 1 device 'vde'. The two clusters are using the same set of nodes. Each node has both vdd and vde as storage node.  

Note:  Some successful case for reference. 
Case 1: one rook-ceph cluster with the same set of storage nodes usinging "vdd" and "vde" .
Case 2: two rook-ceph clusters with two different set of storage nodes each having one device 'vdd'.
Case 3: two rook-ceph clustesr with the same set of storage nodes each having one device 'vdd' and one directory ( /data/rook)

2). The 2nd cluster will have one or two osd pods fails to be brought up randomly. 

=======kubectl -n rook-ceph-system get pods -o wide  ====
NAME                                  READY   STATUS    RESTARTS   AGE   IP              NODE              NOMINATED NODE
rook-ceph-agent-6bh2q                 1/1     Running   0          29m   192.188.88.14   rook-control-01   <none>
rook-ceph-agent-7qbfz                 1/1     Running   0          29m   192.188.88.12   rook-worker-03    <none>
rook-ceph-agent-9dtj9                 1/1     Running   0          29m   192.188.88.8    rook-worker-02    <none>
rook-ceph-agent-q4mp8                 1/1     Running   0          29m   192.188.88.13   rook-control-03   <none>
rook-ceph-agent-rwgw6                 1/1     Running   0          29m   192.188.88.10   rook-control-02   <none>
rook-ceph-agent-t5tv9                 1/1     Running   0          29m   192.188.88.17   rook-worker-01    <none>
rook-ceph-operator-7545bffc9b-z66zd   1/1     Running   0          29m   192.168.1.188   rook-control-01   <none>
rook-discover-7229c                   1/1     Running   0          29m   192.168.1.198   rook-worker-03    <none>
rook-discover-gxvzl                   1/1     Running   0          29m   192.168.1.88    rook-control-03   <none>
rook-discover-l2bm9                   1/1     Running   0          29m   192.168.1.173   rook-worker-01    <none>
rook-discover-p82ft                   1/1     Running   0          29m   192.168.1.16    rook-control-02   <none>
rook-discover-qkfpr                   1/1     Running   0          29m   192.168.1.136   rook-control-01   <none>
rook-discover-vcr2m                   1/1     Running   0          29m   192.168.1.254   rook-worker-02    <none>
=======kubectl -n rook-ceph get pods -o wide  ====
NAME                                             READY   STATUS      RESTARTS   AGE   IP              NODE              NOMINATED NODE
ceph-oam-6dfd5ffccc-bl9cs                        2/2     Running     0          28m   192.188.88.8    rook-worker-02    <none>
rook-ceph-mgr-a-f445598f9-hnhw4                  1/1     Running     0          28m   192.168.1.145   rook-control-01   <none>
rook-ceph-mon-a-5cb7f699c7-6lztx                 1/1     Running     0          28m   192.168.1.134   rook-control-01   <none>
rook-ceph-mon-b-56bd6c759c-xxx7q                 1/1     Running     0          28m   192.168.1.34    rook-control-02   <none>
rook-ceph-mon-c-6c6b779f86-bnh4r                 1/1     Running     0          28m   192.168.1.91    rook-control-03   <none>
rook-ceph-osd-0-5b6dd9f967-knvnz                 1/1     Running     0          27m   192.168.1.176   rook-control-01   <none>
rook-ceph-osd-1-7dd9f79bb7-4q87x                 1/1     Running     0          27m   192.168.1.13    rook-control-02   <none>
rook-ceph-osd-2-967dbf495-knllx                  1/1     Running     0          27m   192.168.1.98    rook-control-03   <none>
rook-ceph-osd-prepare-rook-control-01-jv65h      0/2     Completed   1          27m   192.168.1.148   rook-control-01   <none>
rook-ceph-osd-prepare-rook-control-02-fpxkm      0/2     Completed   1          27m   192.168.1.24    rook-control-02   <none>
rook-ceph-osd-prepare-rook-control-03-xwjvr      0/2     Completed   1          27m   192.168.1.95    rook-control-03   <none>
rook-ceph-rgw-rook-ceph-store-5f684dc766-4k7lt   1/1     Running     0          26m   192.168.1.196   rook-worker-02    <none>
rook-ceph-tools-6b8b666d47-q5lj2                 1/1     Running     0          28m   192.188.88.12   rook-worker-03    <none>
=======kubectl -n rook-ceph2 get pods -o wide  ====
NAME                                          READY   STATUS             RESTARTS   AGE   IP              NODE              NOMINATED NODE
ceph-oam-6dfd5ffccc-67z74                     2/2     Running            0          28m   192.188.88.17   rook-worker-01    <none>
rook-ceph-mgr-a-6b8b9596db-4ms8x              1/1     Running            0          26m   192.168.1.9     rook-control-02   <none>
rook-ceph-mon-a-6669d55bc-v4h6d               1/1     Running            0          26m   192.168.1.132   rook-control-01   <none>
rook-ceph-mon-b-84fcc48df4-hkfvd              1/1     Running            0          26m   192.168.1.35    rook-control-02   <none>
rook-ceph-mon-c-5fbfd69c76-v68bv              1/1     Running            0          26m   192.168.1.97    rook-control-03   <none>
rook-ceph-osd-0-78f9487846-d7qf8              1/1     Running            0          25m   192.168.1.128   rook-control-01   <none>
rook-ceph-osd-1-5d7976f4d6-lzvw4              0/1     CrashLoopBackOff   9          25m   192.168.1.2     rook-control-02   <none>
rook-ceph-osd-2-5646cd948d-czslc              1/1     Running            0          25m   192.168.1.99    rook-control-03   <none>
rook-ceph-osd-prepare-rook-control-01-9g2d2   0/2     Completed          0          25m   192.168.1.155   rook-control-01   <none>
rook-ceph-osd-prepare-rook-control-02-zgfkh   0/2     Completed          0          25m   192.168.1.14    rook-control-02   <none>
rook-ceph-osd-prepare-rook-control-03-k4j5q   0/2     Completed          0          25m   192.168.1.100   rook-control-03   <none>
rook-ceph-tools-6b8b666d47-qthc2              1/1     Running            0          28m   192.188.88.12   rook-worker-03    <none>

3). The rook-ceph2 has one pod failure. And rook-ceph1 is healthy with all 3 pods up. 

ceph -s
  cluster:
    id:     f6f4f98b-8ca2-4817-99d3-be22b0cded16
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum b,c,a
    mgr: a(active)
    osd: 3 osds: 2 up, 2 in
 
  data:
    pools:   0 pools, 0 pgs
    objects: 0  objects, 0 B
    usage:   2.0 GiB used, 8.0 GiB / 10 GiB avail
    pgs:     

ceph osd status
+----+-----------------+-------+-------+--------+---------+--------+---------+------------+
| id |       host      |  used | avail | wr ops | wr data | rd ops | rd data |   state    |
+----+-----------------+-------+-------+--------+---------+--------+---------+------------+
| 0  | rook-control-01 | 1026M | 4090M |    0   |     0   |    0   |     0   | exists,up  |
| 1  |                 |    0  |    0  |    0   |     0   |    0   |     0   | exists,new |
| 2  | rook-control-03 | 1026M | 4090M |    0   |     0   |    0   |     0   | exists,up  |
+----+-----------------+-------+-------+--------+---------+--------+---------+------------+

4). Failure pod log

kubectl -n rook-ceph2 log rook-ceph-osd-1-5d7976f4d6-lzvw4
2019-02-21 15:51:16.059213 I | rookcmd: starting Rook v0.9.2 with arguments '/rook/rook ceph osd start -- --foreground --id 1 --osd-uuid 33615b87-041b-46ef-b0ad-2b8965916930 --conf /var/lib/rook/osd1/rook-ceph2.config --cluster ceph'
2019-02-21 15:51:16.059453 I | rookcmd: flag values: --help=false, --log-level=INFO, --osd-id=1, --osd-store-type=bluestore, --osd-uuid=33615b87-041b-46ef-b0ad-2b8965916930
2019-02-21 15:51:16.059464 I | cephmon: parsing mon endpoints: 
2019-02-21 15:51:16.059469 W | cephmon: ignoring invalid monitor 
2019-02-21 15:51:16.059835 I | exec: Running command: ceph-volume lvm activate --no-systemd --bluestore 1 33615b87-041b-46ef-b0ad-2b8965916930
2019-02-21 15:51:19.683866 I | Running command: /bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-1
2019-02-21 15:51:19.683909 I | Running command: /usr/sbin/restorecon /var/lib/ceph/osd/ceph-1
2019-02-21 15:51:19.683914 I | Running command: /bin/ceph-bluestore-tool --cluster=ceph prime-osd-dir --dev /dev/ceph-a99d9aa9-e356-4fd3-96c8-aade831755f0/osd-data-b5ae0fcd-e483-4169-aa77-a51df8579ad3 --path /var/lib/ceph/osd/ceph-1 --no-mon-config
2019-02-21 15:51:19.683920 I | Running command: /bin/ln -snf /dev/ceph-a99d9aa9-e356-4fd3-96c8-aade831755f0/osd-data-b5ae0fcd-e483-4169-aa77-a51df8579ad3 /var/lib/ceph/osd/ceph-1/block
2019-02-21 15:51:19.683923 I | Running command: /bin/chown -h ceph:ceph /var/lib/ceph/osd/ceph-1/block
2019-02-21 15:51:19.683926 I | Running command: /bin/chown -R ceph:ceph /dev/mapper/ceph--a99d9aa9--e356--4fd3--96c8--aade831755f0-osd--data--b5ae0fcd--e483--4169--aa77--a51df8579ad3
2019-02-21 15:51:19.683929 I | Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-1
2019-02-21 15:51:19.683932 I | --> ceph-volume lvm activate successful for osd ID: 1
2019-02-21 15:51:19.694440 I | exec: Running command: ceph-osd --foreground --id 1 --osd-uuid 33615b87-041b-46ef-b0ad-2b8965916930 --conf /var/lib/rook/osd1/rook-ceph2.config --cluster ceph
2019-02-21 15:51:19.805074 I | failed to fetch mon config (--no-mon-config to skip)
failed to start osd. Failed to complete '': exit status 1. 

5). The logs will be attached. 
[rook-logs02-21.gz](https://github.com/rook/rook/files/2890125/rook-logs02-21.gz)

**Environment**:
* OS (e.g. from /etc/os-release): rhel 7.5
* Kernel (e.g. `uname -a`): Linux rook-control-01 3.10.0-862.14.4.el7.x86_64 #1 SMP Fri Sep 21 09:07:21 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
* Cloud provider or hardware configuration: openstack
* Rook version (use `rook version` inside of a Rook Pod): v0.9.2
* Kubernetes version (use `kubectl version`): v1.12.3
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Tectonic
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox] (https://rook.io/docs/Rook/master/toolbox.html)):

**Comment by yan636** [comment]
I have referenced #2785 about the cleanup. I've the same issue when "hostNetwork" is "false" or "true". Also I've done all the steps on https://rook.io/docs/rook/v0.9/ceph-teardown.html. 

In below case, one of the osds failed to be up. 
-------------------------
kubectl -n rook-ceph2 get pods
NAME                                          READY   STATUS             RESTARTS   AGE
ceph-oam-7ffc6458f8-g7srf                     2/2     Running            0          154m
rook-ceph-mgr-a-6d4659fb97-l6mqs              1/1     Running            0          152m
rook-ceph-mon-a-5b5d6fdf4-kzflg               1/1     Running            0          152m
rook-ceph-mon-b-f5c47dd56-tzbzc               1/1     Running            0          152m
rook-ceph-mon-c-6ddc5b6b55-z2ltv              1/1     Running            0          152m
rook-ceph-osd-0-575c59b68c-hs9df              1/1     Running            0          151m
rook-ceph-osd-1-b6f56d5c7-kgffw               1/1     Running            0          151m
rook-ceph-osd-2-7f7bb9b97f-qm8dl              0/1     CrashLoopBackOff   33         151m
rook-ceph-osd-prepare-rook-control-01-stl6z   0/2     Completed          0          151m
rook-ceph-osd-prepare-rook-control-02-r2jkm   0/2     Completed          0          151m
rook-ceph-osd-prepare-rook-control-03-8qm5n   0/2     Completed          0          151m
rook-ceph-tools-6c594b46b-jv95n               1/1     Running            0          154m
---------------------
The error log:

2019-03-11 16:37:04.933653 I | rookcmd: starting Rook v0.9.3 with arguments '/rook/rook ceph osd start -- --foreground --id 2
--osd-uuid b55ba25f-c267-43e2-a44b-88d8c0fddf08 --conf /var/lib/rook/osd2/rook-ceph2.config --cluster ceph'
2019-03-11 16:37:04.933845 I | rookcmd: flag values: --help=false, --log-level=INFO, --osd-id=2, --osd-store-type=bluestore,
--osd-uuid=b55ba25f-c267-43e2-a44b-88d8c0fddf08
2019-03-11 16:37:04.933856 I | cephmon: parsing mon endpoints:
2019-03-11 16:37:04.933887 W | cephmon: ignoring invalid monitor 
2019-03-11 16:37:04.934218 I | exec: Running command: ceph-volume lvm activate --no-systemd --bluestore 2
b55ba25f-c267-43e2-a44b-88d8c0fddf08
2019-03-11 16:37:08.816169 I | Running command: /bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-2
2019-03-11 16:37:08.816325 I | Running command: /usr/sbin/restorecon /var/lib/ceph/osd/ceph-2
2019-03-11 16:37:08.816442 I | Running command: /bin/ceph-bluestore-tool --cluster=ceph prime-osd-dir --dev
/dev/ceph-6a1fab1a-316c-4929-aed7-51c7962c4d95/osd-data-a9da3233-4fc1-4034-92d4-7749533bf45d --path /var/lib/ceph/osd/ceph-2
--no-mon-config
2019-03-11 16:37:08.816476 I | Running command: /bin/ln -snf
/dev/ceph-6a1fab1a-316c-4929-aed7-51c7962c4d95/osd-data-a9da3233-4fc1-4034-92d4-7749533bf45d /var/lib/ceph/osd/ceph-2/block
2019-03-11 16:37:08.816521 I | Running command: /bin/chown -h ceph:ceph /var/lib/ceph/osd/ceph-2/block
2019-03-11 16:37:08.816547 I | Running command: /bin/chown -R ceph:ceph
/dev/mapper/ceph--6a1fab1a--316c--4929--aed7--51c7962c4d95-osd--data--a9da3233--4fc1--4034--92d4--7749533bf45d
2019-03-11 16:37:08.816573 I | Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-2
2019-03-11 16:37:08.816611 I | --> ceph-volume lvm activate successful for osd ID: 2
2019-03-11 16:37:08.824962 I | exec: Running command: ceph-osd --foreground --id 2 --osd-uuid
b55ba25f-c267-43e2-a44b-88d8c0fddf08 --conf /var/lib/rook/osd2/rook-ceph2.config --cluster ceph
2019-03-11 16:37:08.893397 I | failed to fetch mon config (--no-mon-config to skip)
failed to start osd. Failed to complete '': exit status 1.

**Comment by yan636** [comment]
Has been blocked for several weeks and it's critical for our production deployment.

**Comment by travisn** [comment]
@yan636 Can you share the `storage` section from your cluster CRDs where you're seeing the issue? If you have multiple clusters on the same nodes you will need to be very careful to make sure the devices in the CRD for one cluster do not overlap with the devices in the other cluster. For example, if one cluster uses xvdd, the other cluster cannot use xvdd. Only one cluster can use each device.

**Comment by yan636** [comment]
I've tried several cases for several times. The two clusters have no overlapping at all. Generally, I'm using "/dev/vdc" in the first cluster and "/dev/vdd" in the 2nd cluster on the same set of nodes. The devices are "clean"( as I tried the new volumes ). The odd part is that the failed osd is very random.  Any log or debugging method that I can use to find out more clues.

**Comment by yan636** [comment]
@travisn Appreciate if you can provide some support.

**Comment by yan636** [comment]
I've seen hard-coded "rook_cluster: rook-ceph" in mgr, mon, olsd, rgw deployments.  It's used in both labels and selector section. Please evaluate the impact.

**Comment by travisn** [comment]
@yan636 Are the rook cluster crds in different namespaces? If they are in the same namespace you should see an error. If there is no error in that case, we need to add one. Multiple cluster crds in the same namespace is not supported

**Comment by yanchicago** [comment]
@travisn The rook cluster crds are in different namespace ( rook-ceph2).

**Comment by travisn** [comment]
Per discussion in slack, this error `failed to fetch mon config (--no-mon-config to skip)` can happen when the daemon keyring is incorrect, missing, or otherwise corrupt. If this is hit again, let's check if the keyring in `ceph auth ls` matches what is found in the rook config.

**Comment by chicofranchico** [comment]
I've a similar problem and I realized that two of my OSDs are also missing keys. How do I properly recreate them?

**Comment by yanchicago** [comment]
I can reproduce the issue easily. This time, only osd-2 is crashed with "failed to fetch mon config (--no-mon-config to skip) failed to start osd. Failed to complete '': exit status 1.".

1). kubectl -n rook-ceph2 get pods

NAME                                          READY   STATUS             RESTARTS   AGE
ceph-oam-6888b674bd-94z45                     2/2     Running            0          15m
rook-ceph-mgr-a-7b7d6dbd98-549j4              1/1     Running            0          13m
rook-ceph-mon-a-85b8fcc4f7-cqcbc              1/1     Running            0          14m
rook-ceph-mon-b-848b79654f-b2x8k              1/1     Running            0          14m
rook-ceph-mon-c-69767cdf95-7ss8r              1/1     Running            0          14m
rook-ceph-osd-0-588f9bd665-wnpr4              1/1     Running            0          13m
rook-ceph-osd-1-65656ddf9c-l6btl              1/1     Running            0          13m
rook-ceph-osd-2-5f7494647c-thsmk              0/1     CrashLoopBackOff   7          13m
rook-ceph-osd-prepare-ceph-control-01-vglhs   0/2     Completed          0          13m
rook-ceph-osd-prepare-ceph-control-02-wxrqw   0/2     Completed          0          13m
rook-ceph-osd-prepare-ceph-control-03-lkpb6   0/2     Completed          0          13m

2). ceph auth ls
osd.0
        key: AQDwC8JckR3nOBAA86LZM+ZZvoKcQlT2+y4ZYw==
        caps: [mgr] allow profile osd
        caps: [mon] allow profile osd
        caps: [osd] allow *
osd.1
        key: AQDxC8Jco0r1JhAA6YDOaVS3C9WlhgX8cu3AaQ==
        caps: [mgr] allow profile osd
        caps: [mon] allow profile osd
        caps: [osd] allow *
osd.2
        key: AQDyC8JcwmKuNRAAkHk3WL00x1DiIf1fCUjX7g==
        caps: [mgr] allow profile osd
        caps: [mon] allow profile osd
        caps: [osd] allow *
client.admin
        key: AQCnC8Jcfkv4JhAAzRnkRc11vmwunhOpYSpDkg==
        auid: 0
        caps: [mds] allow *
        caps: [mgr] allow *
        caps: [mon] allow *
        caps: [osd] allow *
client.bootstrap-mds
        key: AQCrC8JcA2RQLBAAHAGicxxk3VlEPTyCB0zRQA==
        caps: [mon] allow profile bootstrap-mds
client.bootstrap-mgr
        key: AQCrC8JcOHVQLBAArUOwGvT7n5qfm4/W26N4nA==
        caps: [mon] allow profile bootstrap-mgr
client.bootstrap-osd
        key: AQCrC8JchYFQLBAARv+tidzQLQ8UxslP+VQFpw==
        caps: [mon] allow profile bootstrap-osd
client.bootstrap-rbd
        key: AQCrC8JcFY1QLBAAfWQxIlg2HDdQsP11gjq8tA==
        caps: [mon] allow profile bootstrap-rbd
client.bootstrap-rgw
        key: AQCrC8Jc3JlQLBAASFeAvne+rAclDpByQ41ZqQ==
        caps: [mon] allow profile bootstrap-rgw
mgr.a
        key: AQDOC8JcmzFEKBAArk+0zVRnOii6tHOnwVMQCw==
        caps: [mds] allow *
        caps: [mon] allow *
        caps: [osd] allow *
installed auth entries:

3). Not sure about how to find it in the ceph config.
cat rook-ceph2.config
[global]
fsid                      = 7d000b1f-af39-4830-99b2-e5fa5b66ace9
run dir                   = /var/lib/rook/osd2
mon initial members       = a b c
mon host                  = 10.254.144.131:6790,10.254.249.193:6790,10.254.125.69:6790
log file                  = /dev/stderr
mon cluster log file      = /dev/stderr
public addr               = 192.168.1.63
cluster addr              = 192.168.1.63
mon keyvaluedb            = rocksdb
mon_allow_pool_delete     = true
mon_max_pg_per_osd        = 1000
debug default             = 0
debug rados               = 0
debug mon                 = 0
debug osd                 = 0
debug bluestore           = 0
debug filestore           = 0
debug journal             = 0
debug leveldb             = 0
filestore_omap_backend    = rocksdb
osd pg bits               = 11
osd pgp bits              = 11
osd pool default size     = 1
osd pool default min size = 1
osd pool default pg num   = 100
osd pool default pgp num  = 100
osd objectstore           = filestore
crush location            = root=default host=ceph-control-03
rbd_default_features      = 3
fatal signal handlers     = false

[osd.2]
keyring              = /var/lib/ceph/osd/ceph-2/keyring
bluestore block path = /var/lib/ceph/osd/ceph-2/block

**Comment by yanchicago** [comment]
I have encountered similar issue in rook 1.0 as well. If there's a guidance/documentation to help to setup 2 clusters, it will be really helpful.

**Comment by travisn** [comment]
To setup two instances of rook on the same K8s cluster the separation would be simplest if the different rook clusters were running mons and osds on a different set of nodes. But if you're using the same nodes, you'll need to be very careful with the following:
- A unique `dataDirHostPath` for each cluster.yaml is required
- A unique set of OSDs must be specified in the cluster.yaml. The operator does not prevent clusters from overwriting each others' OSDs, so the desired state must be very explicit to keep them separate.
   - If you're consuming devices, you need to use a device filter or device list such that the devices will only be used by one of the clusters. 
   - If you're using directories, the directories cannot overlap as they are stored on the host

**Comment by yanchicago** [comment]
Thanks for quick response.  
1). I have used different dataDirHostPath: one is /data0/rook and the other is /data0/rook2
2). Unique set of OSDs:  I have specified storedevices as "vdd" in one cluster and "vdf" for the other cluster.  
Here're cluster.yaml and cluster2.yaml file I used. 

`cat cluster.yaml
---
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  cephVersion:
    image: ceph/ceph:v14.2.1-20190430
    allowUnsupported: false
  dataDirHostPath: /data0/rook
  mon:
    count: 3
    allowMultiplePerNode: False 
  dashboard:
    enabled: True
    port: 8443 
    ssl:  True 
  placement:
    all:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: 'is_control'
              operator: In
              values:
              - "true"
      tolerations:
      - key: 'is_control'
        operator: Exists
  network:
    hostNetwork: False
  rbdMirroring:
    workers: 0 
  resources:
  storage: # cluster level storage configuration and selection
    useAllNodes: True 
    useAllDevices: False
    deviceFilter: 
    devices:
      - name: "vdd"
    location:
    config:
      storeType: bluestore
      metadataDevice: 
      osdsPerDevice: "1" # this value can be overridden at the node or device level

cat cluster2.yaml
---
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph2
  namespace: rook-ceph2
spec:
  cephVersion:
    image: ceph/ceph:v14.2.1-20190430
    allowUnsupported: false
  dataDirHostPath: /data0/rook2
  mon:
    count: 3
    allowMultiplePerNode: False 
  dashboard:
    enabled: True
    port: 8444 
    ssl:  True 
  placement:
    all:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: 'is_control'
              operator: In
              values:
              - "true"
      tolerations:
      - key: 'is_control'
        operator: Exists
  network:
    hostNetwork: False
  rbdMirroring:
    workers: 0 
  resources:
  storage: # cluster level storage configuration and selection
    useAllNodes: True 
    useAllDevices: False
    deviceFilter: 
    devices:
      - name: "vdf"
    location:
    config:
      storeType: bluestore
      metadataDevice: 
      osdsPerDevice: "1" # this value can be overridden at the node or device level

`

**Comment by travisn** [comment]
What OSD pods are created? If the OSDs are being created with ceph-volume, I wonder if there is an issue where both clusters will attempt to start the OSD pods even if they belong to the other cluster. If both clusters are starting the OSDs we would certainly run into issues.

**Comment by yanchicago** [comment]
I have tried to bring up the 2nd cluster after the 1st cluster is up and running. The similar issue would occur for the 2nd cluster. If I restart the operator after the 2nd cluster is deployed, the osd in the 1st cluster sometimes would be crashed. There's conflict between the two clusters. But can't figure out from the log.

**Comment by mkurzydlowski** [comment]
I also am trying to use multiples clusters on a common set of nodes. Every cluster will use it's own device (it will have the same name on every node). But during second cluster installation I see that Ceph volumes are correctly initialized but the newly created OSD sometimes tries to use the wrong Ceph volume.

In fact also the older cluster suffers as some OSD start trying to use the wrong Ceph volume. From what I can tell OSD is chooses the volume which has the last position when sorted alphabetically. For example:

Disk /dev/mapper/ceph--d6033923--1b18--4dac--a79a--98de63409f15-osd--data--f1596ad9--f8e4--44c1--9e2f--a64c4eeaa933: 100 GiB, 107378376704 bytes, 209723392 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

Disk /dev/mapper/ceph--e8d1b0ac--b01d--4ecf--99dd--938a5d783500-osd--data--5de9d646--4b01--4403--93ea--9b7e9bf8c5f0: 100 GiB, 107378376704 bytes, 209723392 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

> 2019-05-23 07:18:56.417255 I | rookcmd: starting Rook v1.0.0 with arguments '/rook/rook ceph osd start -- --foreground --id 0 --osd-uuid 3f2ecc64-6d1c-494c-bd03-bfd75a794858 --conf /var/lib/rook/osd0/irk-wskip.config --cluster ceph'
2019-05-23 07:18:56.417378 I | rookcmd: flag values: --help=false, --log-flush-frequency=5s, --log-level=INFO, --osd-id=0, --osd-store-type=bluestore, --osd-uuid=3f2ecc64-6d1c-494c-bd03-bfd75a794858
2019-05-23 07:18:56.417383 I | op-mon: parsing mon endpoints: 
2019-05-23 07:18:56.417388 W | op-mon: ignoring invalid monitor 
2019-05-23 07:18:56.417576 I | exec: Running command: stdbuf -oL ceph-volume lvm activate --no-systemd --bluestore 0 3f2ecc64-6d1c-494c-bd03-bfd75a794858
2019-05-23 07:18:56.974371 I | Running command: /bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-0
2019-05-23 07:18:57.289701 I | Running command: /usr/sbin/restorecon /var/lib/ceph/osd/ceph-0
2019-05-23 07:18:57.603619 I | Running command: /bin/ceph-bluestore-tool --cluster=ceph prime-osd-dir --dev /dev/ceph-e8d1b0ac-b01d-4ecf-99dd-938a5d783500/osd-data-5de9d646-4b01-4403-93ea-9b7e9bf8c5f0 --path /var/lib/ceph/osd/ceph-0 --no-mon-config
2019-05-23 07:18:58.001482 I | Running command: /bin/ln -snf /dev/ceph-e8d1b0ac-b01d-4ecf-99dd-938a5d783500/osd-data-5de9d646-4b01-4403-93ea-9b7e9bf8c5f0 /var/lib/ceph/osd/ceph-0/block
2019-05-23 07:18:58.296671 I | Running command: /bin/chown -h ceph:ceph /var/lib/ceph/osd/ceph-0/block
2019-05-23 07:18:58.601807 I | Running command: /bin/chown -R ceph:ceph /dev/mapper/ceph--e8d1b0ac--b01d--4ecf--99dd--938a5d783500-osd--data--5de9d646--4b01--4403--93ea--9b7e9bf8c5f0
2019-05-23 07:18:58.877758 I | Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
2019-05-23 07:18:59.187904 I | --&gt; ceph-volume lvm activate successful for osd ID: 0
2019-05-23 07:18:59.201094 I | exec: Running command: ceph-osd --foreground --id 0 --osd-uuid 3f2ecc64-6d1c-494c-bd03-bfd75a794858 --conf /var/lib/rook/osd0/irk-wskip.config --cluster ceph
2019-05-23 07:18:59.283031 I | failed to fetch mon config (--no-mon-config to skip)
failed to start osd. Failed to complete '': exit status 1.

@travisn wrote:
> What OSD pods are created? If the OSDs are being created with ceph-volume, I wonder if there is an issue where both clusters will attempt to start the OSD pods even if they belong to the other cluster. If both clusters are starting the OSDs we would certainly run into issues."

What do you mean by "both clusters are starting the OSDs"? Certainly both clusters run their own OSDs (pods) but I don't see any other way that would work. Maybe I misunderstood how running multiple clusters works?

**Comment by travisn** [comment]
Ok, I see the bug in the osd creation. The sequence for OSD creation is the following:
- 

The cluster fsid is returned by `ceph-volume list` so rook needs to match it to the current cluster fsid instead of assuming it already belongs to the cluster. 
```
$ ceph-volume lvm list --format json
...
            "tags": {
                "ceph.cluster_fsid": "aeec10f6-b350-4ad3-a670-486c0d45feed", 
                ...
             }
```

The [getCephVolumeOSDs](https://github.com/rook/rook/blob/master/pkg/daemon/ceph/osd/volume.go#L191) method should only return the OSDs that correspond to the local cluster ID. Until this issue is fixed, the only workaround for running multiple ceph clusters is to have the different clusters consume devices on different nodes.

**Comment by sp98** [comment]
@yan636 @yanchicago Need some help to reproduce it.  Can you please share the common.yaml, operator.yaml files that you are using. Thanks.

**Comment by yanchicago** [comment]
The cluster.yaml is listed earlier. Attached is the operator.yaml that is installed through helm chart. cluster_common.yaml and cluster_common2.yaml are also attached. 


[2cluster.zip](https://github.com/rook/rook/files/3228057/2cluster.zip)

**Comment by yanchicago** [comment]
When will this fix be available? Will this be part of 1.0.x?

**Comment by travisn** [comment]
Yes, this is being backported and will be out in 1.0.2 hopefully in a couple days. I'll reopen to track the backport.

**Comment by travisn** [comment]
#3266 is merged, so this will be included with 1.0.2...
